Docs/event payload scoping spec (#3539)
conveyordata/data-product-portal: docs/adr/reference/use-case-aws-s3-provisioner.md

### Technical Asset Link Approval

A technical asset can be linked to an output port. Linking takes effect immediately — there is no separate approval step for the TA↔OP relationship. The TA's own PENDING / ACTIVE / ARCHIVED status still controls whether it can be linked at all (only ACTIVE TAs can be linked).

---

## Data Product Events

### Data Product Created

The provisioner needs the data product's namespace, domain, type, and the list of initial team members (with their roles).

1. Resolve the domain bucket name from the domain (by convention: `{domain-slug}`).
2. Create the IAM role `dpp-{namespace}`:
   - Trust policy: include all initial team members as principals immediately. Creation always includes at least one owner; do not wait for `team_member_added` events to populate the trust policy.
   - Permission policy: read/write access to `s3://{domain-bucket}/{namespace}/*`.
   - If the data product type confers additional permissions (e.g. cross-account access for a shared type), apply them now.
3. Create the S3 prefix by uploading a placeholder or relying on the first write.

> 4. Call the Portal API to register the subfolder as a `default` technical asset on this data product. This triggers a `technical_asset.created` event — that handler will update the IAM role. Make the handler idempotent so re-running it on an already-existing prefix is safe.

Reviewer: ```suggestion
4. Call the Portal API to register the subfolder as a `default` technical asset on this data product. This triggers a `technical_asset.created` event. That handler should be idempotent and see we already have access.
```

Author: ?  `and see we already have access.` Not clear what that means. Clarify

Author: and see if the policy is already up-to-date.

5. Update the data product status from `Draft` to `Ready`.

> **Note**: The data product type may influence the base permissions attached to the IAM role. The exact mapping is platform-specific and should be documented per type. Tags and status could similarly influence permissions, but their semantics are not yet defined.

---

### Data Product Updated

> The provisioner needs the data product's namespace, the old and new domain, the new type, and the list of current technical assets (with their S3 configuration).

Reviewer: ```suggestion
The provisioner needs the data product's namespace, the new domain, the new type, and the list of current technical assets (with their S3 configuration).
```

Author: ?

Reviewer: You do not need the old domain, you can always figure this out.

If the s3 path you expect does not exist on the location you can check the other domains to see if they had that s3 path, and kick of the migration.

And alternative is that we always send the before state of your whole data product.

Only type and domain changes have infrastructure impact.

**Type changed:**

1. Update the IAM role `dpp-{namespace}` to reflect the permission set associated with the new type.

**Domain changed:**

> This is the most disruptive change. Proceed carefully:

Reviewer: ```suggestion
This is the most disruptive change. To figure out we are in this case, we need to check the bucket of the new domain, if there is no folder for our current data product we know we need to check all other domains to see if our data product has a folder there. If they do we know we need to migrate it
```

Author: You suggest we scan all buckets for the previous location instead of getting the old value?

1. Identify the new domain bucket from the new domain name.
2. Call the Portal API to create a new `default` technical asset pointing to `s3://{new-domain-bucket}/{namespace}/`. Wait for the `technical_asset.created` event to confirm the new prefix is provisioned and the IAM role has access.
3. Copy all data from `s3://{old-domain-bucket}/{namespace}/` to `s3://{new-domain-bucket}/{namespace}/`.
4. For each output port that had the old technical asset linked:
   a. Link the new technical asset to that output port.
   b. For each data product with an approved input port on that output port, grant read access to the new S3 location on their IAM role.
5. Remove read/write access to the old S3 prefix from `dpp-{namespace}`.
6. For each consumer data product identified in step 4b, revoke their access to the old S3 location.

> 7. Mark the old technical asset as archived via the Portal API.

Reviewer: I don't know if this is something you should support as a team.

An alternative is that we do not copy the new data, or delete the old prefix. But that that is manual, we could send out an email or something instead with the procedure?

The copy could also take hours or days.

Author: This will need to happen at some customers regardeless so I rather have it documented.

> > **Warning**: This sequence involves multiple Portal API calls that each trigger further events. Design all handlers to be idempotent.

Reviewer: You might need to put your events in a queue, so that new events for this data product can't be handled while there are ongoing changes.

Author: I would add a warning on the challenges rather then listing a solution.

Reviewer 2: The challenge here will be more eventual consistency than just idempotency no?

Author: Not to self: Update remark that we should think about concurrent updates.

---

### Data Product Deleted

> **Important**: Deletion cascades in the database. Technical assets, output ports, and input port links are all deleted in the same transaction without emitting individual events for each. The provisioner must handle all teardown in this single event.

The provisioner needs the data product's namespace, its settings, the full S3 configuration of each technical asset, and — critically — for each technical asset, the list of consumer data products that had approved read access via any output port exposing that asset. This consumer list is gone from the database by the time the event fires, so it must be available at event dispatch time.

1. For each technical asset in the payload:
   a. Revoke read access to this TA's S3 location from the IAM role of every consumer data product that had approved access.
   b. If `read-only-role` is enabled on any of those consumers, also revoke from `dpp-{consumer-namespace}-ro`.
2. If the `read-only-role` setting was enabled on this data product: delete `dpp-{namespace}-ro`.
3. Delete the IAM role `dpp-{namespace}`.
4. Archive or delete S3 objects under `s3://{domain-bucket}/{namespace}/`.

> > **Recovery note**: Restoring a deleted data product requires both S3 object recovery (possible via bucket versioning) and database recovery. Enable versioning on all domain buckets. There is no automated restore path — restoring the data product record requires a database recovery operation.

Reviewer: ```suggestion
> **Recovery note**: Restoring a deleted data product requires both S3 object recovery (possible via bucket versioning). Or moving it to a trash folder on s3 that cleans up files after 30 days.
There is no automated restore path — restoring the data product record requires a manual s3 copy or restoring of versioned files
```

Author: I made the mistake of solutiong. We should not do this in this document. Why remove database recovery?

---

### Data Product Settings Changed

The provisioner needs to know which setting changed, the data product's namespace, and the full list of its technical assets (to set up the read-only role's permissions) and approved input port links (to grant the read-only role access to consumed data as well). Since the data product still exists at this point, a round-trip to fetch this is acceptable — but knowing at minimum which setting changed and the namespace is required before any round-trip.

**`read-only-role` enabled:**

1. Create IAM role `dpp-{namespace}-ro`.
2. Grant read-only access to all S3 locations of the data product's own technical assets (`s3:GetObject`, `s3:ListBucket` on each `s3://{bucket}/{prefix}/*`).
3. Grant read-only access to all S3 locations of every approved input port (i.e. all TAs the data product already has approved access to via its input ports).
4. Apply the same trust policy as the main role so all current team members can also assume the read-only role.

**`read-only-role` disabled:**

1. Delete `dpp-{namespace}-ro`.

> > Warn any external consumers (e.g. BI tools) that were granted access to this role before deleting it.

Reviewer: Should we take into account how to warn them? We can send out an email to users who have access?

Author: Part of Portal or Provisioner? I had the feeling Katarina wants to add these things into Portal natively.

Author: Remove this.

**`scheduled` disabled:**

1. Remove the MWAA-related permissions from `dpp-{namespace}`.

---

### Team Member Added

Both `owner` and `developer` have integration access. The provisioner needs the data product's namespace, the user's email (to resolve their AWS principal), and their role.

1. Resolve the user's AWS principal (IAM user ARN or SSO identity) from their email.
2. Add the principal to the trust policy of `dpp-{namespace}`.
3. If `read-only-role` is enabled: also add the principal to the trust policy of `dpp-{namespace}-ro`.

---

### Team Member Removed

The provisioner needs the data product's namespace, the user's email, and their role.

1. Resolve the user's AWS principal from their email.
2. Remove the principal from the trust policy of `dpp-{namespace}`.
3. If `read-only-role` is enabled: also remove the principal from the trust policy of `dpp-{namespace}-ro`.

---

### Team Member Updated (role changed)

Since both `owner` and `developer` have integration access, a role change does not require any change to the IAM trust policy. No action required.

> > **Note**: If future roles are introduced without integration access, the provisioner would need the user's email, their old role, and their new role to decide whether to add or remove the principal from the trust policy. The old role is not available when the update event fires — it would need to be tracked externally or included alongside the new role.

Reviewer: ```suggestion
> **Note**: If future roles are introduced without integration access, the provisioner would need the user's email, and their new role. They can then read the roles of the data product to decided whether to add or remove the principal from the trust policy. 
The old role is not needed since the provisioner can just check all roles
```

If link status is **APPROVED**:

1. For each technical asset linked to the output port, grant read access (`s3:GetObject`, `s3:ListBucket`) on `dpp-{requesting-namespace}`.
2. If `read-only-role` is enabled on the requesting data product: apply the same grants to `dpp-{requesting-namespace}-ro`.

If link status is **PENDING**: no action. Wait for `output_port.link_approved`.

---

### Input Port Request Approved

Fires when the output port owner approves a pending access request.

The provisioner needs the requesting data product's namespace and settings, and the S3 configuration of the technical assets behind the output port.

1. For each technical asset linked to the output port, grant read access on `dpp-{requesting-namespace}`.
2. If `read-only-role` is enabled on the requesting data product: apply the same grants to `dpp-{requesting-namespace}-ro`.

---

### Input Port Request Denied

No action required. The link was PENDING; no access was ever granted.

---

### Input Port Removed

Fires when a data product's access to an output port is removed. This can happen to both approved and pending links.

> The provisioner needs the requesting data product's namespace and settings, the link status at the time of removal (APPROVED or PENDING), and the S3 configuration of the technical assets behind the output port. Without the status the provisioner cannot tell whether access was ever granted and therefore whether there is anything to revoke.

Reviewer: ```suggestion
The provisioner needs the requesting data product's namespace and settings, the link status at the time of removal (APPROVED or PENDING), and the S3 configuration of the technical assets behind the output port. Without the status the provisioner would need to check the AWS to tell whether access was ever granted and therefore whether there is anything to revoke.
```

If the link was **APPROVED**:

1. For each technical asset that was linked to the output port, check whether the requesting data product still has access to the same TA via another approved input port.
2. For any TA no longer accessible via another path: revoke read access from `dpp-{requesting-namespace}`.
3. If `read-only-role` is enabled: also revoke from `dpp-{requesting-namespace}-ro`.

If the link was **PENDING**: no action. No access was ever granted.

---

## Exploration Events

Explorations are lightweight workspaces owned by a single user. Unlike data products they have no output ports, no technical assets, and no settings. Their sole purpose is to consume data from approved input ports.

The provisioner creates an IAM role for the exploration so the owner can assume it to access S3 locations they have been granted read access to via input ports. No S3 prefix is created — explorations do not produce or store data.

### Exploration Created

The provisioner needs the exploration's namespace and the owner's identity.

1. Create IAM role `dpp-{namespace}` with the owner as the sole principal in the trust policy.
2. Update the exploration status from `Draft` to `Ready`.

> > Explorations do not own storage — no S3 prefix or technical asset is created.

Reviewer 2: Don't they require some sort of scratch space that is not shared?

Reviewer: I do believe there are use cases where that makes sense indeed, but ok for me to just make things a bit easier and ignore it 😄

Author: Indeed, would be similar to a DP anyway.

If the link was **PENDING**: no action.

---

## Output Port Events

### Output Port Created

No action required.

---

### Output Port Updated

No action required.

---

### Output Port Deleted

> **Important**: Technical assets are owned by the data product, not by the output port. Deleting an output port does **not** delete the technical assets. It removes the links between the output port and its TAs. The data product retains all its technical assets.

The provisioner needs the S3 configuration of the technical assets that were linked to this output port, and the list of consumer data products (with their namespace and settings) that had approved input port access. Both are gone from the database after deletion, so they must be available at event dispatch time.

1. For each consumer data product with an approved input port on this output port:
   a. For each TA that was linked to this output port, check whether the consumer still has access to the same TA via another approved input port.
   b. For any TA no longer accessible via another path: revoke read access from `dpp-{consumer-namespace}`.

>    c. If `read-only-role` is enabled on the consumer: also revoke from `dpp-{consumer-namespace}-ro`.

Reviewer: Maybe we should just not allow users to delete output ports without first removing all consumers?

Author: Maybe, maybe not. Would leave it documented for now until we support this.

---

### Output Port Team Member Added / Removed / Updated

Output ports have their own role assignments — users can be granted roles such as approver (the ability to approve or deny access requests to the output port). These are Portal-level roles with no AWS infrastructure equivalent.

No infrastructure action required.

> **Note**: Future platform requirements could introduce output-port-level IAM grants — for example, granting a specific user direct S3 read access to the output port's technical assets without going through a data product input port. If that is introduced, these events would be the trigger. For now they are informational only.

---

## Technical Asset Events

### Technical Asset Created

The provisioner needs the owning data product's namespace, settings, the technical mapping (default or custom), and the S3 configuration (bucket, prefix or location).

**Default (portal-managed subfolder):**

1. Create the S3 prefix (upload a placeholder object if needed).

> 2. Grant read/write access to `s3://{bucket}/{prefix}/*` on `dpp-{namespace}`.

Reviewer: What is the relation between this and the `s3://{domain}/{namespace}` location we created before? 

Maybe we don't need that standard space for data products?

Author: I made a design decision for the platform. You get a default space. You can register subfolders in case that is what you want to share instead of the root folder.  I think stap 2 needs to be removed as we should already have this kind of access. 

This was done deliberate as our customers platforms aren't perfect either. I think the design still holds.

Author: Needs to be domein namespace

   b. If `read-only-role` is enabled on the consumer: also grant on `dpp-{consumer-namespace}-ro`.

---

### Technical Asset Unlinked from Output Port

Fires when a technical asset is removed from an output port.

The provisioner needs the S3 configuration of the technical asset and the list of consumer data products (with their namespace and settings) that had approved input port access to this output port at the time of removal.

1. For each consumer data product that had approved input port access to this output port:
   a. Check whether the consumer still has access to this TA via another output port.
   b. If not: revoke read access from `dpp-{consumer-namespace}`.
   c. If `read-only-role` is enabled on the consumer: also revoke from `dpp-{consumer-namespace}-ro`.

---

## Information Requirements Summary

This table captures information the provisioner needs that is not trivially available at event dispatch time. It is intended as input for event design.

| Event | Information needed | Why |
|---|---|---|
| `data_product.created` | Initial team members with their roles | Trust policy must be populated at creation, not via subsequent events |

> | `data_product.deleted` | Per-TA: list of consumers with approved access | Object is already gone; provisioner must revoke consumer access |

Reviewer: We maybe should just require that first all consumers must be removed? That you can't delete a data product with consumers, I think this makes sense

> | `data_product.input_port_linked` | Link status (APPROVED / PENDING) | Provisioner must distinguish unrestricted (act now) from restricted (wait) |
> | `data_product.input_port_unlinked` | Link status at removal time | Provisioner must know whether access was ever granted |

Reviewer: ```suggestion
| `data_product.input_port_unlinked` | Link status at removal time | Provisioner must know whether access was ever granted or read the status of AWS |
```

Author: Why remove linked?
